ansible-vault: prevent discarding the vault file ownership (#11544)

eprigorodov opened the pull request:
##### SUMMARY

`ansible-vault` operations (encrypt/decrypt/edit) replace processed vault file with its new version. Running `ansible-vault` by another user changes file ownership and can make vault inaccessible for the original owner.

This PR replaces file shuffling with write to the original file.

Fixes #11544

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
ansible-vault

##### ANSIBLE VERSION
```
ansible 2.6.0 (bugfix-11544-vault-owner 1c546a6d76) last updated 2018/05/15 18:25:56 (GMT +200)
  config file = None
  configured module search path = [u'/home/<user>/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/<user>/dev/ansible/lib/ansible
  executable location = /home/<user>/dev/ansible/bin/ansible
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```

##### ADDITIONAL INFORMATION

Problem description and bugfix rationale are captured in #11544.

eprigorodov committed:
ansible-vault: prevent discarding the vault file ownership (#11544)

Fixes: #11544

Do not replace original vault file during encrypt/decrypt/edit
operations, just overwrite old inode with the new content.

Add unit tests checking that vault operations do not replace vault file.

eprigorodov commented:
For external contributor it is hardly possible to fix the [failed test in Shippable](https://app.shippable.com/github/ansible/ansible/runs/65512/51/tests), because it is reported as failed due to a network timeout:

> Failures
> 
> download_binary_modules / /root/ansible/test/integration/targets/binary_modules/download_binary_modules.yml:15 / [testhost_binary_modules] testhost_binary_modules: download binary module url=https://ansible-ci-files.s3.amazonaws.com/test/integration/roles/test_binary_modules/{{ module_filename }}, dest={{ playbook_dir }}/library/{{ module_filename }}, mode=493
> 
> failure: Request failed: <urlopen error timed out>
> 
> {
> "changed": false, 
> "dest": "/root/ansible/test/integration/targets/binary_modules/library/helloworld_freebsd", 
> "msg": "Request failed: <urlopen error timed out>", 
> "state": "absent", 
> "url": "https://ansible-ci-files.s3.amazonaws.com/test/integration/roles/test_binary_modules/helloworld_freebsd"
> }
> 

How can one re-run the check?

mkrizek commented:
@eprigorodov I just restarted the CI job.

resmo reviewed (approved):
LGTM

eprigorodov commented:
Is it possible to help the review somehow? Let me help: the actual change touches 25 lines, everything else is four new unit tests. 

`VaultEditor._shred_file()` will accept a new parameter, `remove=True`, which allows to shred file in-place without deleting it (similar to `wipe -k`).

New parameter is used in the `VaultEditor.write_file()`, so it does not remove original file before writing its new content.

The behaviour of `VaultEditor._edit_file_helper()` changed. Before the PR, it wrote changed and re-encrypted data into the temporary file, then called `VaultEditor.shuffle_files()`. The latter deleted original vault, renamed temporary file and then tried to restore its permissions and ownership. That failed for non-root users.

PR changes `VaultEditor._edit_file_helper()` to make it write changed data directly into the original inode. That behaviour is identical to what `VaultEditor.rekey_file()` already does.

The method `VaultEditor.shuffle_files()` becomes obsolete as well as `chmod()` / `chown()` calls, so they are removed.

Unit tests invoke encrypt, decrypt, edit and rekey operations and then check that processed vault file has the same inode number, ownership and permissions.

bcoca commented on lib/ansible/parsing/vault/__init__.py:
```diff
@@ -878,10 +879,8 @@ def _edit_file_helper(self, filename, secret,
         # so decode to unicode here
         b_ciphertext = self.vault.encrypt(b_tmpdata, secret, vault_id=vault_id)
-        self.write_data(b_ciphertext, tmp_path)
+        self.write_data(b_ciphertext, filename)
```
This creates a concurrency issue, now the file is unsafe to use as someone might be editing it while someone else is trying to use it in a play. The shuffle relies on 'atomic file system operations' which would prevent any user from seeing a incomplete/mid update file.

To do this change safely you must introduce locking to both the editing and the reading of vault files.

eprigorodov replied:
Thank you for pointing out, indeed there is a chance of a read/write that can conflict with concurrent and incomplete operation.

I would like to try adding a thin locking layer. It will be based on the following observation: all vault operations first read its contents into the memory using `VaultEditor.read_data()`, then process it in the memory, then write results back using `VaultEditor.write_data()`. The edit operation does the same cycle twice: decrypting into temporary plaintext file, then encrypting back into the original vault file.

Thus locking will need to be introduced between `read_data()` and `write_data()` methods. It will not try to lock the vault file itself due to platform compatibility difficulties. It will operate via additional lock file in vim style (e.g. `.vault_filename.lock`). It will make sure that lock file is deleted on error via `atexit` handler. An unsuccessful locking attempt will wait for 5 seconds and then fail globally due to a timeout.

Please tell if you see any problems with this approach. Please also tell if the branch should be rebased on top of the current head.

bcoca replied:
I have many issues with lockfiles, for one `atexit` cannot guarantee their disappearance, only on clean shutdown of the program.

You also need to lock when ansible/ansible-console/ansible-playbook read the files for runtime use, or they can get a incomplete data stream, which is why we use rename, which is atomic and ensures either old or new are read and no 'in between' state is accessible.

eprigorodov replied:
You express two concerns: 1) that particular locking approach is difficult to implement properly and 2) that, even being implemented in `ansible-vault`, locking still cannot ensure data consistency.

The second concern seems to be more important, so let us model the worst case. `ansible` / `ansible-console` / `ansible-playbook` do not write to the vault, so the fail scenario seems to be the following:

* vault functionality has been designed to support keeping playbooks in version control systems,
* with VCS the normal cycle is: change the vault file, commit the playbook, release, checkout new playbook version onto the controller host, execute,
* despite this design the user changes vault file directly on the controller host, while running the playbook that uses it,
* the vault file is large enough for giving significant chance for concurrent reads,
* `ansible-playbook` reads incomplete data and fails.

Considering, that the current implementation raises an error when run by non-root user (recorded in the issue), I think that this kind of failure can be neglected.

If you'd agree then we can discuss more reliable design of locking.

If you have already evaluated such trade-offs and have concluded that the renaming approach is the best possible then please tell. I understand your work load and don't insist on arguing.

bcoca replied:
I would not dismiss any scenario, too many times have i encountered something like you describe (as well as the complete opposite) for every point.

I'm fine that we add locking (i would make it an option, not the default), I just want it to be added properly as otherwise you end up  with a false sense of security and a race condition waiting to happen. Your setup might not break, but we need to consider much more than that.

I did evaluate trade-offs, which is why we generally use rename instead of locks. If using an SCM to keep the file, permission changes would be apparent in the commit, also i would expect the deployment to ensure correct permissions on the target. 

But, as above, that is not always the case, which is why I'm open to adding the feature, to support other setups.

eprigorodov replied:
Thank you, I understand it as a blessing to try adding locking. It should be free of races, should work on NFS mounts, should not hang on stale lock files and does not need to support running `ansible-vault` on Windows. Is that correct?

And could you please explain a bit better idea of "add locking, making it an option, not the default"? 

The only purpose of locking was to avoid the loss of ownership, which can only happen when user does not own the vault file. Do you mean that `ansible-vault` should support both renaming and locking schemes and only fall back to locking when that is appropriate?

bcoca replied:
or via `--exclusive` switch.

eprigorodov replied:
The requested change has been implemented in e6585de08e3d8ba49a4fd1fbed76d555d828a3d2 + 9da37a0c0c0765628dee6a4d3c06bf6dcf38a1e2.

eprigorodov committed:
ansible-vault: protect concurrent edits with locking (#11544)

Restore rename()-based concurrency protection in _edit_file_helper().
Add locking to ensure consistent vault modification and reading.
Add unit tests for the new locking class.
Add '--exclusive' command line parameter to switch from rename()-based
to lock-based protection.

eprigorodov committed:
ansible-vault: fix CI findings (#11544)

The lack of permission does not make VaultEditorLock.acquire() fail when run by root.
Fix Python 3 compatibility issue in test_vault_lock_cooperation().
Fix an error raised by locking non-existing vault file.

eprigorodov commented:
@bcoca, @mkrizek: Both unsuccessful checks in the last run (72013) failed when trying to allocate a virtual machine in the AWS. That cannot be fixed by changing the code, could you please help?

In all other respects the change looks working now.

eprigorodov commented:
Thanks to someone, all checks have passed. @bcoca, could you please have a look if the change looks safe enough now?

sebasptsch commented:
Is there any update on this issue?

eprigorodov commented:
Well, I can rebase and re-request review if anyone would be willing to have a look.

eprigorodov commented:
I'd like to retract this PR as the original #11544 is closed. Thank you for your time, Ansible developers.
